src/libraries/System.Net.WebSockets.Client/src/System/Net/WebSockets/HttpResponseHeadersReadOnlyCollection.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Net.Http.Headers;
+
+namespace System.Net.WebSockets
+{
+    internal sealed class HttpResponseHeadersReadOnlyCollection : IReadOnlyDictionary<string, IEnumerable<string>>
+    {
+        private readonly Dictionary<string, IEnumerable<string>> _headers = new Dictionary<string, IEnumerable<string>>();
+
+        public HttpResponseHeadersReadOnlyCollection(HttpResponseHeaders headers)
+        {
+            foreach (KeyValuePair<string, IEnumerable<string>> header in headers)
+            {
+                _headers.Add(header.Key, header.Value);
+            }
+        }
+
+        public IEnumerable<string> this[string key] => _headers[key];
+
+        public IEnumerable<string> Keys => _headers.Keys;
+
+        public IEnumerable<IEnumerable<string>> Values => _headers.Values;
+
+        public int Count => _headers.Count;
+
+        public bool ContainsKey(string key) => _headers.ContainsKey(key);
+        public IEnumerator<KeyValuePair<string, IEnumerable<string>>> GetEnumerator() => _headers.GetEnumerator();
+        public bool TryGetValue(string key, [MaybeNullWhen(false)] out IEnumerable<string> value) => _headers.TryGetValue(key, out value);
+        IEnumerator IEnumerable.GetEnumerator() => _headers.GetEnumerator();
```

Comment:
Shouldn't HttpResponseHeadersReadOnlyCollection just store a reference to the HttpResponseHeaders instance rather than making a copy, and just have all of the members access that underlying collection (ideally through the NonValidated view)?  Otherwise this is much more expensive than it would seem it needs to be.

Reply:
Ok, it seems to be just one option among others, I'm going to replace it with a reference

Reply:
I updated collection to use reference and thought, would it be ok in case of exception, when the response message is going to be disposed?

Reply:
It should be fine.  We're relying on an implementation detail, but this is also all in netcoreapp (plus it's extremely likely other folks are relying on that implementation detail such that we'd never change it).